shell/server/core.js

```diff
@@ -0,0 +1,220 @@
+    } else if (notification.userId !== Meteor.userId()) {
+      throw new Meteor.Error(403, "Notification does not belong to current user.");
+    } else {
+      dismissNotification(notificationId);
+    }
+  },
+  readAllNotifications: function () {
+    if (!Meteor.userId()) {
+      throw new Meteor.Error(403, "User not logged in.");
+    }
+    Notifications.update({userId: Meteor.userId()}, {$set: {isUnread: false}}, {multi: true});
+  }
+});
+
+NotificationHandle.prototype.close = function () {
+  var self = this;
+  return inMeteor(function () {
+    tryDropNotification(self.notificationId);
+  });
+};
+
+NotificationHandle.prototype.save = function () {
+  var self = this;
+  return inMeteor(function () {
+    var sturdyRef = new Buffer(Random.id(20));
+    var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+    ApiTokens.insert({
+      _id: hashedSturdyRef,
+      frontendRef: {
+        notificationHandle: self.notificationId
+      }
+    });
+    return {sturdyRef: sturdyRef};
+  });
+};
+
+SandstormCoreImpl.prototype.restore = function (sturdyRef) {
+  var self = this;
+  return inMeteor(function () {
+    var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+    var token = ApiTokens.findOne(hashedSturdyRef);
+    if (!token) {
+      throw new Error("No token found to restore");
+    }
+    if (token.frontendRef) {
+      var notificationId = token.frontendRef.notificationHandle;
+      return {cap: makeNotificationHandle(notificationId)};
+    } else if (token.objectId) {
+      return waitPromise(openGrain(self.grainId).supervisor.restore(token.objectId).catch(function (err) {
+        if (shouldRestartGrain(err, 0)) {
+          return openGrain(grainId, true).supervisor.restore(token.objectId);
+        }
+      }));
+    } else {
+      throw new Error("Unknown token type.");
+    }
+  });
+};
+
+var dropToken = function (grainId, sturdyRef) {
+  var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+  var token = ApiTokens.findOne({_id: hashedSturdyRef});
+  if (!token) {
+    return;
+  }
+  if (token.frontendRef) {
+    var notificationId = token.frontendRef.notificationHandle;
+    ApiTokens.remove({_id: hashedSturdyRef});
+    tryDropNotification(notificationId);
+  } else if (token.objectId) {
+    dropWakelock(grainId, token.objectId.wakeLockNotification);
+  } else {
+    throw new Error("Unknown token type.");
+  }
+};
+
+SandstormCoreImpl.prototype.drop = function (sturdyRef) {
+  var self = this;
+  return inMeteor(function () {
+    dropToken(self.grainId, sturdyRef);
+  });
+};
+
+SandstormCoreImpl.prototype.makeToken = function (ref, owner) {
+  var self = this;
+  return inMeteor(function () {
+    var sturdyRef = new Buffer(Random.id(20));
+    var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+    // TODO(soon): should userId be filled?
```

Comment:
This relates to the TODO in the doc comment for the RPC method. What `userId` is really meant to track is the edge in the sharing graph leading to this cap. In the case of background processing, there is none. Eventually `makeToken()` should take another parameter which can be filled in to indicate a parent user or a parent ApiToken.